nexus/src/app/background/tasks/instance_watcher.rs

```diff
@@ -436,23 +436,11 @@ impl BackgroundTask for InstanceWatcher {
         opctx: &'a OpContext,
     ) -> BoxFuture<'a, serde_json::Value> {
         async {
-            let mut tasks = tokio::task::JoinSet::new();
-            // We're using the size of the database query page as a simple
-            // mechanism for limiting the number of concurrent health checks: if
-            // we only query for `MAX_CONCURRENT_CHECKS` records at a time, and
-            // we wait until all spawned health checks have completed before
-            // reading the next batch, it follows that there are only ever
-            // `MAX_CONCURRENT_CHECKS` checks in flight at a time.
-            //
-            // This does mean that, unlike using a counting semaphore or
-            // similar, we may have fewer than the limit health checks running
-            // in parallel as we will not query for a new batch until *all*
-            // checks for the current batch have completed.  If that becomes an
-            // issue, we could implement a more sophisticated
-            // concurrency-limiting scheme later.
-            let mut paginator = Paginator::new(MAX_CONCURRENT_CHECKS);
-
-            let mut total: usize = 0;
+            let mut paginator = Paginator::new(nexus_db_queries::db::datastore::SQL_BATCH_SIZE);
+            let mut tasks = ParallelTaskSet::new_with_parallelism(
```

Comment:
As mentioned in chats - mostly posting this for visibility to others - one small issue with the `ParallelTaskSet` as it exists today is that it'll grow in memory proportional to the number of spawns, even if the actual amount of parallel task work is limited to `MAX_CONCURRENT_CHECKS`.

So this does mean we'd end up spawning "a task per all instances" before joining them.

At our current scale, that's *probably* okay, but it may not be in the future.

Reply:
So @smklein and I spent the last three hours noodling on this and I am pleased to report that it is now more or less impossible to have more than the concurrency limit number of tasks resident in memory when using the new API!